### PR TITLE
Test: collect and upload

### PR DESCRIPTION
**Pull request by gremat**
Depends on #3.

### CODE REVIEW

File: go/README.md [hunk 1]
```diff
@@ -11,6 +11,9 @@ Usage: amd64-windows-thunderstorm-collector.exe [OPTION]...
       --ca strings                   Path to a PEM CA certificate that signed the HTTPS certificate of the Thunderstorm server.
                                      Specify multiple CAs by using this flag multiple times.
       --debug                        Print debugging information.
+      --exclude strings              Paths that should be excluded. Supports globbing with ?, * and **.
```
Comment (secDre4mer) [inline go/README.md]: Changes seem to overlap with the other PR, is this intended? Should we process the other one first?

File: go/collector_test.go [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+		{
+			"with excludes",
+			CollectorConfig{
+				RootPaths:    []string{testRoot},
+				ExcludeGlobs: []string{"**/sub?", "**/*.jpg"},
+			},
+			[]filedata{
+				{"foo.txt", []byte("foo\n")},
+			},
+			CollectionStatistics{1, 1, 2, 0, 0},
```
Comment (secDre4mer) [inline go/collector_test.go]: This is not particularly readable - maybe swap to named field initialization

File: go/collector_test.go [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+	numFound := 0
+	for _, collected := range collectedFiles {
+		for _, expected := range expectedFilesFound {
+			if collected.path == filepath.Join(testRoot, expected) {
+				numFound++
+			}
+		}
+	}
+	if numFound != len(expectedFilesFound) {
```
Comment (secDre4mer) [inline go/collector_test.go:258]: This checks that all expected files are uploaded; it does not, however, guarantee that only the expected files are uploaded

File: go/collector_test.go [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+			if diff := cmp.Diff(tc.expectedStats, stats, cmp.Exporter(func(_ reflect.Type) bool { return true })); diff != "" {
+				t.Fatalf("Statistics mismatch: %s", diff)
+			}
+			for _, expected := range tc.expectedFilesFound {
+				t.Run(fmt.Sprintf("File %s", expected.path), func(t *testing.T) {
+					found := false
+					expectedPathAbs, _ := filepath.Abs(filepath.Join(testRoot, expected.path))
+					for _, received := range receivedFiles {
+						if received.path == filepath.Base(expected.path) ||
+						(filepath.IsAbs(received.path) && received.path == expectedPathAbs) { // FIXME: Starting with Go 1.17.0, filepath.Walk consistently reports relative paths when called with relative paths.
```
Comment (secDre4mer) [inline go/collector_test.go]: Couldn't this be fixed more easily by collecting from `filepath.Abs("testdata")`?

Alternatively: if this causes information loss from relative paths, the collector should probably call `filepath.Abs` itself on the root path.

**Review by secDre4mer — CHANGES_REQUESTED** [review]

File: go/README.md (hunk 1 above)
Comment (gremat) [reply go/README.md]: Yes, see top-level comment/PR description.

File: go/collector_test.go (hunk 2 above)
Comment (gremat) [reply go/collector_test.go]: I wanted to have it concise and easily repeatable. But you're right: initialized by naming the fields might be more reader-friendly. Changed.

File: go/collector_test.go (hunk 3 above)
Comment (gremat) [reply go/collector_test.go:258]: Added another check.

File: go/collector_test.go (hunk 4 above)
Comment (gremat) [reply go/collector_test.go]: Initially, I did not want to use absolute paths in the check, especially when the root paths are provided as relative paths. But I had another look into the handling of the root paths from the beginning to the end now and I realized two things: `getFileContentAsFormData()` already uses _absolute_ paths so there is no point in sticking to relative paths. And there was a problem with extracting the correct filepath from the multipart data. I have now fixed the latter and I use absolute paths for the checks.

**Review by secDre4mer — DISMISSED** [review]

**Review by secDre4mer — APPROVED** [review]

**Pull request merged**